src/mail/mail.service.ts

```diff
@@ -0,0 +1,100 @@
+import { Injectable } from "@nestjs/common";
+import axios, { AxiosResponse } from "axios";
+import * as ejs from "ejs";
+
+@Injectable()
+export class MailService {
+  private readonly mailgunAPIBaseURL = `${process.env.MAILGUN_API_URL}/${process.env.MAILGUN_EMAIL_DOMAIN}`;
+  private readonly emailFromString = `${process.env.MAILGUN_DEFAULT_NAME} <${process.env.MAILGUN_DEFAULT_EMAIL}>`;
+
+  // Private helper function to send a payload to mailgun API.
+  // Exists to reduce code redundancy.
+  private sendMailPayload(params: Record<string, any>): Promise<AxiosResponse> {
+    return axios.post(
+      `${this.mailgunAPIBaseURL}/messages`,
+      {} /* Mailgun needs the payload to be in params, leave body empty */,
+      {
+        auth: { username: "api", password: process.env.MAILGUN_API_KEY },
+        params
+      }
+    );
+  }
+
+  // Sends a simple email with a plain text body.
+  async sendTextEmail(
+    emailTo: string,
+    subject: string,
+    bodyText: string,
+    emailsCC?: [string],
+    emailsBCC?: [string]
+  ): Promise<[number, string]> {
+    const payload = {
+      from: this.emailFromString,
+      to: emailTo,
+      cc: emailsCC ? emailsCC.join(",") : undefined,
+      bcc: emailsCC ? emailsBCC.join(",") : undefined,
+      subject,
+      text: bodyText
+    };
+
+    try {
+      const res = await this.sendMailPayload(payload);
+      return [res.status, "Email queued."];
+    } catch (err) {
+      return [err.status, "Error in sending email."];
+    }
+  }
+
+  // Sends a simple email with an HTML body.
+  async sendHTMLEmail(
+    emailTo: string,
+    subject: string,
+    bodyHTML: string,
+    emailsCC?: [string],
+    emailsBCC?: [string]
+  ): Promise<[number, string]> {
+    const payload = {
+      from: this.emailFromString,
+      to: emailTo,
+      cc: emailsCC ? emailsCC.join(",") : undefined,
+      bcc: emailsCC ? emailsBCC.join(",") : undefined,
+      subject,
+      html: bodyHTML
+    };
+
+    try {
+      const res = await this.sendMailPayload(payload);
+      return [res.status, "Email queued."];
+    } catch (err) {
+      return [err.status, "Error in sending email."];
+    }
+  }
+
+  // Sends an email to a **single** recipient after rendering the given template.
+  async sendTemplatedEmail(
+    emailTo: string,
+    subject: string,
+    templatePath: string,
+    templateVariables: Record<string, any>,
+    emailsCC?: [string],
+    emailsBCC?: [string]
+  ): Promise<[number, string]> {
```

Comment:
- Same as the comment above, let's use an object instead of a bunch of arguments.
- templatePath should just maybe be templateFile where we manually add the `./template/${templateFile}`

Reply:
Should all possible params be combines into one DTO? Function specific params like `bodyText`, `bodyHTML` or template params could be optional and be checked by the function being called?

Or would it be better to use separate DTOs for each function?